Python/Product/Analysis/LanguageServer/OpenFiles.cs

```diff
@@ -0,0 +1,115 @@
+        private readonly ConcurrentDictionary<Uri, OpenFile> _files = new ConcurrentDictionary<Uri, OpenFile>();
+        private readonly ILogger _log;
+        private readonly ProjectFiles _projectFiles;
+
+        public OpenFiles(ProjectFiles projectFiles, ILogger log) {
+            _projectFiles = projectFiles;
+            _log = log;
+        }
+        public OpenFile GetDocument(Uri uri) => _files.GetOrAdd(uri, _ => new OpenFile(_projectFiles, _log));
+        public void Remove(Uri uri) => _files.TryRemove(uri, out _);
+    }
+
+    sealed class OpenFile {
+        private readonly ILogger _log;
+        private readonly ProjectFiles _projectFiles;
+
+        public IDictionary<int, BufferVersion> LastReportedDiagnostics { get; } = new Dictionary<int, BufferVersion>();
+        public List<DidChangeTextDocumentParams> PendingChanges { get; } = new List<DidChangeTextDocumentParams>();
+
+        public OpenFile(ProjectFiles projectFiles, ILogger log) {
+            _projectFiles = projectFiles;
+            _log = log;
+        }
+
+        public void DidChangeTextDocument(DidChangeTextDocumentParams @params, Action<IDocument> enqueueAction) {
+            var changes = @params.contentChanges;
+            if (changes == null) {
+                return;
+            }
+
+            var uri = @params.textDocument.uri;
+            var doc = _projectFiles.GetEntry(uri) as IDocument;
+            if (doc == null) {
+                return;
+            }
+
+            try {
+                var part = _projectFiles.GetPart(uri);
+                _log.TraceMessage($"Received changes for {uri}");
+
+                var docVersion = Math.Max(doc.GetDocumentVersion(part), 0);
+                var fromVersion = Math.Max(@params.textDocument.version - 1 ?? docVersion, 0);
+
+                if (fromVersion > docVersion && @params.contentChanges?.Any(c => c.range == null) != true) {
+                    // Expected from version hasn't been seen yet, and there are no resets in this
+                    // change, so enqueue it for later.
+                    _log.TraceMessage($"Deferring changes for {uri} until version {fromVersion} is seen");
+                    lock (PendingChanges) {
+                        PendingChanges.Add(@params);
+                    }
+                    return;
+                }
+
+                var toVersion = @params.textDocument.version ?? (fromVersion + changes.Length);
+
+                doc.UpdateDocument(part, new DocumentChangeSet(
+                    fromVersion,
+                    toVersion,
+                    changes.Select(c => new DocumentChange {
+                        ReplacedSpan = c.range.GetValueOrDefault(),
+                        WholeBuffer = !c.range.HasValue,
+                        InsertedText = c.text
+                    })
+                ));
+
+                DidChangeTextDocumentParams? next = null;
+                lock (PendingChanges) {
```

Comment:
`lock` on publicly available object

Reply:
That's the way it used to be. Create separate lock object?

Reply:
Yep.

Reply:
Does it even need to be publicly visible? It should only be relevant within this handler.

Reply:
Made private